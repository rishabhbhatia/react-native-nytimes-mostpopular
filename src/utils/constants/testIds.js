/* -------------------------------------------------------------------------- */
/*                  testIDS of application are declared here                  */
/* -------------------------------------------------------------------------- */

/* ------------------------------ articles screen ------------------------------ */
export const TEST_ID_ARTICLES_SCREEN_LIST = "articlesScreenList";
export const TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW = "articleDetailsScreenWebview";
