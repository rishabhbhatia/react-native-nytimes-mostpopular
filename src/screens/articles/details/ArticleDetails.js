import React from "react";
import {
  SafeAreaView, View, Text, ActivityIndicator,
} from "react-native";

import { WebView } from "react-native-webview";

import styles from "./styles";
import { TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW } from "../../../utils/constants/testIds";

/* -------------------------------------------------------------------------- */
/*                 Screen to display details of an article                    */
/* -------------------------------------------------------------------------- */
export default class ArticleDetails extends React.PureComponent {
  /* ------------------------ custom header properties ------------------------ */
  static navigationOptions = {
    title: "Article Details",
    headerStyle: {
      backgroundColor: "#47e4c2",
    },
    headerTintColor: "#fff",
  };

  constructor(props) {
    super(props);
    this.state = {
      articleUrl: "",
    };
  }

  componentDidMount() {
    const articleUrl = this.getArticleUrlFromProps();
    this.setArticleUrlInState(articleUrl);
  }

  /* --------------- returns article url passed from navigation --------------- */
  getArticleUrlFromProps = () => {
    const { navigation } = this.props;
    return navigation.getParam("articleUrl", "");
  }

  /* --------------- updates the state with value of article url -------------- */
  setArticleUrlInState = (articleUrl) => {
    this.setState({
      articleUrl,
    });
  }

  /* ------------------- displayed while article is loading ------------------- */
  renderLoading = () => (
    <View style={styles.container}>
      <ActivityIndicator size="large" />
    </View>
  )

  /* ------------------ displayed when article fails to load ------------------ */
  renderError = () => (
    <View style={styles.container}>
      <Text>Failed to load article, try again later.</Text>
    </View>
  )

  render() {
    const { articleUrl } = this.state;
    return (
      <SafeAreaView style={styles.container}>
        <WebView
          testID={TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW}
          source={{ uri: articleUrl }}
          startInLoadingState
          renderLoading={this.renderLoading}
          renderError={this.renderError}
        />
      </SafeAreaView>
    );
  }
}
