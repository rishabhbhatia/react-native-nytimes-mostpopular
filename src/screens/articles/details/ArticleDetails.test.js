import React from "react";
import { shallow } from "enzyme";
import ArticleDetails from "./ArticleDetails";
import { TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW } from "../../../utils/constants/testIds";


const navigation = { getParam: () => { return { articleUrl: "https://www.nytimes.com/interactive/2019/07/03/us/us-citizenship-test-quiz.html " } } };

/* -------------------------------------------------------------------------- */
/*                                  UI Tests                                  */
/* -------------------------------------------------------------------------- */
describe("ArticleDetails", () => {
    /* -------------------------------- snapshot -------------------------------- */
    it("matches snapshot", () => {
        expect.assertions(1);
        const articleDetailsScreen = shallow(<ArticleDetails navigation={navigation} />);
        expect(articleDetailsScreen).toMatchSnapshot();
    });

    it("article details view is visible", () => {
        expect.assertions(1);
        const articleDetailsScreen = shallow(<ArticleDetails navigation={navigation} />);
        const articleDetailsWebview = articleDetailsScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW);
        expect(articleDetailsWebview.exists()).toBeTruthy();
    });

    it("article details loading renders correctly", () => {
        expect.assertions(1);
        const articleDetailsScreen = shallow(<ArticleDetails navigation={navigation} />);
        const articleDetailsWebview = articleDetailsScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW);
        expect(articleDetailsWebview.props().renderLoading()).toBeTruthy();
    });

    it("article details error renders correctly", () => {
        expect.assertions(1);
        const articleDetailsScreen = shallow(<ArticleDetails navigation={navigation} />);
        const articleDetailsWebview = articleDetailsScreen.findWhere(node => node.props().testID === TEST_ID_ARTICLE_DETAILS_SCREEN_WEBVIEW);
        expect(articleDetailsWebview.props().renderError()).toBeTruthy();
    });
});
