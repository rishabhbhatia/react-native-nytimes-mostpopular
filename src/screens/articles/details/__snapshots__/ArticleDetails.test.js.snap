// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleDetails matches snapshot 1`] = `
<SafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <WebView
    cacheEnabled={true}
    originWhitelist={
      Array [
        "http://*",
        "https://*",
      ]
    }
    renderError={[Function]}
    renderLoading={[Function]}
    source={
      Object {
        "uri": Object {
          "articleUrl": "https://www.nytimes.com/interactive/2019/07/03/us/us-citizenship-test-quiz.html ",
        },
      }
    }
    startInLoadingState={true}
    testID="articleDetailsScreenWebview"
    useSharedProcessPool={true}
    useWebKit={true}
  />
</SafeAreaView>
`;
